New SafeBoot website from @soylentOrange

mathieucarbou opened the pull request:
New SafeBoot website from @soylentOrange.

Duplicate of PR #12, but editable.

mathieucarbou commented:
@soylentOrange : I took your commits in a branch and added one for some cleanups.
I also updated MycilaESPConnect and added the `-D ESPCONNECT_NO_LOGGING`.
I also tried to simplify the ota mode, but still not enough.

soylentOrange commented:
@mathieucarbou Well, seems that working in parallel messed up some things. Even though I considered adding the last hostname to ESPConnect might be beneficial, I didn't include it in my PR to it (and only updated the logging code). I've added a something to Safeboot to read the preferences of a safeboot section for the last known hostname. 

Let me try to get my additions in working order with your update to ESPConnect...

mathieucarbou commented:
@soylentOrange : here is a status:

- MycilaESPConnect is done (macro + hostname)
- Integrated it in main
- Also updated CI to add html-minifier-terser
- Grabbed your commits
- Added one at the top to fix some stufff

But as you see in the build, it is not passing. Formware size is still too big with the new UI

soylentOrange commented:
I modified the update.html as I missed something to remove.

```
Successfully created esp32 image.
safeboot([".pio/build/esp32dev/firmware.bin"], [".pio/build/esp32dev/firmware.elf"])
safeboot.py: Firmware size valid: 655344 <= 655360
```
And there is the luxury to spend additional 16 bytes...

But, my problem is that I don't quite understand the -U option in platformio. When setting in the general env:
```
; ------------------------------
; Logging OFF by default
; -D MYCILA_LOGGER_SUPPORT
-D ESPCONNECT_NO_LOGGING
; ------------------------------
```

and trying to enable logging for ESPConnect as well in one of the specific boards:
```
[env:esp32-c3-devkitc-02]
board = esp32-c3-devkitc-02
build_flags =
  ${env.build_flags}
  -D MYCILA_SAFEBOOT_FANCY_WEBSITE
  -U ESPCONNECT_NO_LOGGING
  -D MYCILA_LOGGER_SUPPORT
extra_scripts = ${env.extra_scripts}
  pre:tools/prepare_html.py
board_build.embed_files = assets/minified_update.html.gz
```

will result in an compilation error:
```
Building in release mode
Compiling .pio/build/esp32-c3-devkitc-02/__compiled_constants.c.o
Converting .pio/build/esp32-c3-devkitc-02/minified_update.html.gz.txt.o
<command-line>: error: macro names must be identifiers
<command-line>: warning: extra tokens at end of #undef directive
```

Do you have an idea what I did wrong?

btw. not having set ESPCONNECT_NO_LOGGING in the general env (and not having the undefine in [env:esp32-c3-devkitc-02] compiles perfectly fine.

mathieucarbou commented:
- `-D ESPCONNECT_NO_LOGGING` will disable any logging from ESPConnect lib
- `-D MYCILA_LOGGER_SUPPORT` will activate logging only for safeboot. Note that we could have just renamed `MYCILA_LOGGER_SUPPORT` to `SAFEBOOT_LOGGING` it would have been easier to understand ;-) We went the complicated route lol

Maybe try a `git fetch upsteam` and `git reset --hard upstream/pr-12` : I have forced-pushed quite a lot to fix little thing also in main and was rebasing, sorry... I will stop and append commits now instead.

mathieucarbou commented:
@soylentOrange : everything is working fine. This is so awesome! I find the UI even better then ElegantOTA. It is simple and efficient. 

Thats a very cool contribution to the project!

I will merge and release now.

soylentOrange commented:
On no, I was too slow. 
I actually updated the logging logic fine grained enabling MYCILA_LOGGER_SUPPORT independently from ESPCONNECT_LOGGING and SAFEBOOT_LOGGING.
But I thinks it's a nice and shiny release now.

mathieucarbou commented:
> On no, I was too slow. I actually updated the logging logic fine grained enabling MYCILA_LOGGER_SUPPORT independently from ESPCONNECT_LOGGING and SAFEBOOT_LOGGING. But I thinks it's a nice and shiny release now.

Well, I will be glad to merge other improvements if you send them :-)
Yes that is a nice release indeed!

soylentOrange commented:
> About MYCILA_LOGGER_SUPPORT, there is only ESPConnects logs that would eventually maybe matter, but I really think we can silent them. Only having some logs from Safeboot image to follow startup is more than enough. So we do not need `MYCILA_LOGGER_SUPPORT`, which is more relevant when using several libs I did.

I agree (actually having no logging from a working SafeBoot version would be fine for me as well).

I'm unsure if it is more relevant to have logging enabled for further boards or having the option to upload a file system image. I played around with the code, after a few modifications to reduce size in the website and the esp code, and removing the option to upload into the file system, logging can be enabled on the boards:
- esp32-c6-devkitc-1
- esp32-poe-iso
- esp32-poe
- wt32-eth01

Especially for some of the eth-enabled boards logging could be useful to get the hostname/ip-address of the devices.
@mathieucarbou I you think it's worth it, I will issue a PR for this change of feature set.

mathieucarbou commented:
> Especially for some of the eth-enabled boards logging could be useful to get the hostname/ip-address of the devices.
> @mathieucarbou I you think it's worth it, I will issue a PR for this change of feature set.

I finally like the idea of being able to update the FS. 
I know several projects relying on FS for web pages and who would use such feature.

The UI as it is right now, IMO, is really perfect! It is beautiful and allows updating for both Firmware and FS. You did a terible job! That was precisely the thing missing in this project!

As for logging, this is not critical, but if the user is troubleshooting something going wrong, it can help him assert that the safeboot feature is working fine, and help him reach the ESP.

But I would not rely on logging to reach the ESP because if the user is able to see the logs, it means he is connected to the ESP somehow, so in that case he does not need safeboot to do his update ;-) Whether on WiFi or Eth. There are so many tools allowing a discovery on the network...

What would be really nice is to be able to have mDNS on every board though... 
I don't know if some lighter mDNS implementations exist, or if directly using the esp-idf API could improve flash usage. 
I didn't check yet with the PIO memory analysis tool is the mdns big object comes from Arduino core or ESP-IDF.

soylentOrange commented:
Ok, thanks for letting me know your mind on this issue (and also for the warm words).
If I happen to find a more lightweight alternative for mDNS, I will give it a try...

mathieucarbou commented:
Arduino WebServer is 12kb on flash... That's a lot for what is needed here... If we can find a little webserver implemetation, i am pretty sure mDNS would fit!

mathieucarbou commented:
@soylentOrange : tried to replace Arduino WebServer by ESP-IDF http_server: this is worse.
https://github.com/mathieucarbou/MycilaSafeBoot/pull/14

mathieucarbou commented:
Also tried with ESPAsync: https://github.com/mathieucarbou/MycilaSafeBoot/pull/15... No luck...
